src/shared/tpm2-util.h

```diff
@@ -259,6 +269,173 @@ int tpm2_tpm2b_public_to_fingerprint(const TPM2B_PUBLIC *public, void **ret_fing
+                                });                                     \
+        })
+
+/* Similar to tpm2_marshal() but uses greedy_realloc() to append the marshalled data. The buf must be usable
+ * with greedy_realloc(). The value of buf (i.e. location of allocated memory) may be modified, so it must be
+ * an lvalue. Returns 0 if successful or < 0 on an error. */
+#define tpm2_marshal_realloc(desc, src, buf, offset, ret_size)          \
+        (tpm2_dlopen() ?: _tpm2_marshal_realloc(desc, src, buf, offset, ret_size, UNIQ))
+#define _tpm2_marshal_realloc(desc, src, buf, offset, ret_size, u)      \
+        ({                                                              \
+                const char *UNIQ_T(DESC, u) = (desc);                   \
+                typeof(src) UNIQ_T(SRC, u) = (src);                     \
+                void **UNIQ_T(BUF, u) = (void**) &(buf);                \
+                size_t UNIQ_T(O, u) = (offset);                         \
+                size_t UNIQ_T(S, u);                                    \
+                _tpm2_marshal_size(UNIQ_T(DESC, u), UNIQ_T(SRC, u), &UNIQ_T(S, u), CONCATENATE(u, _tpm2_marshal_realloc)) \
+                        ? /* != 0 (failure) */                          \
+                        : greedy_realloc(UNIQ_T(BUF, u), UNIQ_T(O, u) + UNIQ_T(S, u), 1) \
+                        ? _tpm2_marshal(UNIQ_T(DESC, u), UNIQ_T(SRC, u), *UNIQ_T(BUF, u), UNIQ_T(O, u) + UNIQ_T(S, u), UNIQ_T(O, u), (ret_size), CONCATENATE(u, _tpm2_marshal_realloc)) \
+                        : log_oom();                                    \
+        })
+
+/* Unmarshal data from buf, starting at offset, into dst. The size of buf is max. If successful and ret_size
+ * is not NULL, it is set to offset plus the number of unmarshalled bytes. Returns 0 on success or < 0 on
+ * error. */
+#define tpm2_unmarshal(desc, dst, buf, max, offset, ret_size)           \
+        (tpm2_dlopen() ?: _tpm2_unmarshal(desc, dst, buf, max, offset, ret_size, UNIQ))
+#define _tpm2_unmarshal(desc, dst, buf, max, offset, ret_size, u)       \
+        ({                                                              \
+                const char *UNIQ_T(DESC, u) = (desc);                   \
+                log_debug("Unmarshalling %s", UNIQ_T(DESC, u));         \
+                size_t UNIQ_T(O, u) = (offset);                         \
+                TSS2_RC UNIQ_T(RC, u) = _UNMARSHAL(dst)(buf, max, &UNIQ_T(O, u), dst); \
+                UNIQ_T(RC, u) == TSS2_RC_SUCCESS     \
+                        ? _tpm2_marshalling_update_ret_size(UNIQ_T(O, u), (ret_size), CONCATENATE(u, _tpm2_unmarshal)) \
+                        : log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), \
+                                          "Failed to unmarshal %s: %s", \
+                                          desc, sym_Tss2_RC_Decode(UNIQ_T(RC, u))); \
+        })
```

Comment:
hmpff, these "functions" are conceptually low-level "library-like" basics and we said that "library-like" functions should not log on their own. but here you do...

Reply:
I wonder if it wouldn't be better to use functions instead of macros here, since the macros are very "function-like" and unnecessarily cryptic.